crates/code_generation/src/perception_databases.rs

```diff
@@ -0,0 +1,94 @@
+use proc_macro2::TokenStream;
+use quote::quote;
+use source_analyzer::cyclers::{CyclerKind, Cyclers};
+
+fn generate_perception_updates(cyclers: &Cyclers) -> TokenStream {
+    let updates_fields = cyclers.instances_with(CyclerKind::Perception).map(
+        |(cycler, instance)| {
+            let field_name = &instance.name;
+            let cycler_module_name = &cycler.name;
+            quote! {
+                pub #field_name: framework::Update<crate::structs::#cycler_module_name::MainOutputs>
+            }
+        },
+    );
+    let mut timestamp_array_items = cyclers
+        .instances_with(CyclerKind::Perception)
+        .map(|(_cycler, instance)| {
+            let field_name = &instance.name;
+            quote! {
+                self.#field_name.first_timestamp_of_non_finalized_database
+            }
+        })
+        .peekable();
+    let find_min_timestamp = if timestamp_array_items.peek().is_some() {
```

Comment:
Since there is a difference between minimum and minimal we should not abbreviate:

```suggestion
    let find_minimum_timestamp = if timestamp_array_items.peek().is_some() {
```

Reply:
Renamed to `find_earliest_timestamp` to make the meaning clearer.
The fact that we search for the smallest number is just a detail, we really want the earliest timestamp.